lib/derivation.ml

```diff
@@ -96,5 +96,35 @@ let rec derive context = function
         Subtype (tyvar_mfa, TyAny)] (* cannot be TyFun (.., ..) because the arity is unknown. give up *)
      in
      Ok (tyvar_mfa, Conj cs)
-  | other ->
-     Error (Printf.sprintf "unsupported type: %s" (show_expr other))
+  | Case (e, clauses) ->
+    (* translate pattern to expression *)
+    let rec pattern_to_expr = function
+      | PatVar v -> Var v
+      | PatTuple es -> Tuple (es |> List.map ~f:(fun e -> pattern_to_expr e)) in
+    (* Var(p) *)
+    let rec variables = function
+      | PatVar v -> [(v, new_tyvar ())]
+      | PatTuple es -> es |> List.map ~f:(fun e -> variables e)
+                          |> List.fold_left ~init:[] ~f:(fun a b -> List.append a b) in
+    derive context e >>= fun (ty_e_t, c_e) ->
+    let beta = new_tyvar () in
+    let rs = clauses |> result_map_m ~f:(fun ((p_n, g_n), b_n) ->
+      let p_n_tyvars = variables p_n in
+      let p_n_expr = pattern_to_expr p_n in
+      (* A ∪ { ... } *)
+      let added_context = 
+        List.fold_left
+        ~f:(fun context (v, tyvar) -> Context.add (Context.Key.Var v) tyvar context)
+        ~init:context
+        p_n_tyvars in
+      (* |- p_n : alpha_n, c_p *)
+      derive added_context p_n_expr >>= fun(ty_alpha_n, c_p) ->
+      (* PAT *)
+      derive added_context g_n >>= fun(ty_g_n, c_g) ->
+      (* |- b_n : beta_n *)
+      derive added_context b_n >>= fun(ty_beta_n, c_b) ->
+        Ok (ty_beta_n, Conj [Eq (beta, ty_beta_n); Eq (ty_e_t, ty_alpha_n); c_p; c_g; c_b])
+    ) in
```

Comment:
`rs` is a list of `typ` and `constraint_` . But the typs do not seem to be used.
If typs does not need, `rs` may be a list of constraints only.

Reply:
It's fixed in the commit https://github.com/kmizu/fialyzer/commit/1c71c939192038499190f58087ac89910dd55f0a